Proposed snapshot should contain only seen transactions

abailly-iohk opened the pull request:
The motivation for this work comes from the idea of packing multiple transactions in a ReqTx (see #900) to guarantee the atomicity of a sequence of txs. Thinking about this feature, I realised that:

1. To enforce atomicity, we would need the snapshot to contain not only individual tx but actually the ReqTx, plus the proposed new UTxO set
2. Our current head logic says nothing about the txs part of a snapshot beyond that they should be applicable to the current ledger.
3. It's therefore impossible to implement the feature and more importantly it's super easy for any node to front-run other nodes: When it's your turn to make a snapshot, just look at the proposed txs pick and chose the one you wants, and add more applicable txs of your liking. In the context of a market, or an auction perhaps, this makes it trivial to cheat as far as I understand it.

Therefore, I initially thought we should augment the protocol to (at least) require that txs in a snapshot have been previously seen in a `ReqTx`. But this is in essence what #728 implies as this implies the only way to properly apply a transaction from the `ReqSn` propagating only TxId is to have seen it before and be able to resolve it. 

Therefore, the goal of this PR is to fix #728 


---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

abailly committed:
Add test checking snapshot contains seen txs

This test is currently failing on purpose. The motivation for this
work comes from the idea of packing multiple transactions in a ReqTx
to guarantee the atomicity of a sequence of txs. Thinking about this
feature, I realised that:

1. To enforce atomicity, we would need the snapshot to contain not
only individual tx but actually the ReqTx, plus the proposed new UTxO
set
2. Our current head logic says nothing about the txs part of a
snapshot beyond that they should be applicable to the current ledger.
3. It's therefore impossible to implement the feature and more
importantly it's super easy for any node to front-run other nodes:
When it's your turn to make a snapshot, just look at the proposed txs
pick and chose the one you wants, and add more applicable txs of your
liking. In the context of a market, or an auction perhaps, this makes
it trivial to cheat as far as I understand it.

Therefore, I think we should augment the protocol to (at least)
require that txs in a snapshot have been previously seen in a ReqTx.

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -144,11 +145,17 @@ spec = do
+        let s0 = inOpenState threeParties ledger
+            reqSn = NetworkEvent defaultTTL $ ReqSn alice 1 [SimpleTx 1 mempty (utxoRef 1)]
+        update bobEnv ledger s0 reqSn
+          `shouldBe` Wait (WaitOnSeenTx 1)
```
This would be exactly the behavior if we would not submit transactions with the `ReqSn`  as actually specified. Do you think this is different in the overall semantics seen by the users?

ch1bo reviewed (commented):
> Therefore, I think we should augment the protocol to (at least) require that txs in a snapshot have been previously seen in a ReqTx.

In the original protocol a `ReqSn` was only ever requesting transactions to be snapshotted by id. We (over-)simplified this to have snapshot requests include the transactions directly. This is definitely a performance problem, but maybe you are on to something bigger here? The issue https://github.com/input-output-hk/hydra/issues/728 would tackle this.

> it's super easy for any node to front-run other nodes

How? Any node would apply the transaction once seen to their seen ledger and not agree with another transaction spending the same input, but differently. Or is this maybe a use case where the thing to be "raced on" is not represented as a UTxO being spent?

ghost commented:
The problem I am seeing is not related to performance, only to the fact that a node can receive a snapshot containing arbitrary transactions, even ones it has not yet seen, as long of course as that snapshot makes sense with the latest known one. This makes it possible for a node to "front-run" another node by: 1/ receiving a `ReqTx tx` that some other party posted then 2/ craft a `ReqSn` not containing this `tx` but another `tx'` that "front-runs" `tx`, for example by consuming some UTxO with a smart contract for swaps, or a bid for auction, or something else. 

Perhaps this is not a real problem? As explained in the PR's description this dawned on me while trying to think about how a node could propose "atomic" transaction sets and how to enforce that in the protocol.

ch1bo commented:
> Perhaps this is not a real problem?

No, I think you are right and this is a real problem. Which makes #728 much more important!

pgrange commented:
Imagine B want to be evil and sign snapshot with its own not submitted transaction instead of another legit one. Suppose it's B's turn to be the leader. I understand that the concern of this issue would be to prevent the following scenario to occur:

1. A submits transaction Ta
2. B ignores Ta
3. B ask for the signature of a new snapshot including transaction Tb not yet submitted
4. A signs the snapshot

If we implement this issue, now A will only sign the snapshot if it has seen Tb before receive the snapshot. I can see the following scenario leading to the same result as before:

1. A submits transaction Ta
2. B ignores Ta
3. B submit transaction Tb
4. B ask for the signature of a new snapshot including transaction Tb just submitted
5. A signs the snapshot

Are these scenario correct in the context of this issue or did I miss something?

ghost commented:
@pgrange You are right, this issue exposes only part of the problem of the lack of fairness in the Hydra consensus protocol. But a big difference with the previous situation is that if `A` has seen both `Ta` and `Tb` this means they are both compatible with the lates confirmed snapshot so it guarantees that when `A` turns to sign snapshot comes, it will be able to include `Ta`.

abailly committed:
Add check for snapshot txs to be in the seen set

abailly committed:
Update specification with seen txs requirement

abailly committed:
Update CHANGELOG

abailly committed:
Remove TODO and replace with spec reference

ch1bo commented on spec/offchain.tex:
```diff
@@ -171,9 +171,9 @@ \subsubsection{Processing transactions off-chain}
-% have all the transaction submitted and needs to be applicable
+snapshot is confirmed ($\bars = \hats$), all requested transactions
+$\mT_{res}$ are applicable to $\barmU$, and all requested transactions have
+been seen ($\mT_{res} \subseteq \hatmT$).
```
There is also the `spec/fig_offchain_prot.tex` with the composite pseudo code which needs updating.

ch1bo replied:
This needs slight updating as we should now mention that "requested transaction ids have been seen and the corresponding transactions are applicable"

ch1bo reviewed (changes requested):
I think this is a protocol change in the direction where we actually want to be (as also described in #728).

Not sure if it's worthwhile to stop half-way and not do the full switch into `ReqSn` only holds txId. That would be more intuitive, I guess, as then the resolution from `TxId -> Tx` is naturally implying the check we do here manually now.

I can approve this change, but would be more comfortable if we would have researchers also think about this (they proposed the only send TxId with lookup of TxId -> Tx in the past, so it should be fine as this is semantically the same though).

Requested change is about updating figure 10 in the spec accordingly.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -659,26 +665,31 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxs =
-                }
-          )
-          [NetworkEffect $ AckSn party snapshotSignature sn]
+      -- Spec: wait T ⊆ \hatT
```
```suggestion
      -- Spec: wait T ⊆ T̂
```

Not sure if unicode is better than latex here though (I used unicode below)

ch1bo replied:
I was using "Combining" unicode, this one is
![image](https://github.com/input-output-hk/hydra/assets/2621189/a82aa5f9-9d74-4487-86f7-14d5c820d3e7)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -691,6 +702,13 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxs =
+  notSeenTxs = fromList (txId <$> requestedTxs) `Set.difference` fromList (txId <$> seenTxs)
+
+  waitSeenTxs continue =
+    if null notSeenTxs
```
Could: write this more in line to the math notation using [`isSubsetOf`](https://hackage.haskell.org/package/containers-0.6.7/docs/Data-Set.html#v:isSubsetOf)

 ```suggestion
     if fromList requestedTxs `isSubsetOf` fromList seenTxs
```

ghost commented:
After discussing this with @ch1bo it seems to make more sense to bite the bullet and implement the "proper" `ReqSn`, eg. send only transaction ids

abailly committed:
Add missing ToCBOR/FromCBOR instances for TxId

abailly committed:
Use TxId instead of Tx when emitting ReqSn

abailly committed:
Update off-chain spec for 'new' condition for ReqSn

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -370,15 +371,20 @@ deriving instance (IsTx tx, IsChainState tx) => FromJSON (Outcome tx)
   = WaitOnNotApplicableTx {validationError :: ValidationError}
   | WaitOnSnapshotNumber {waitingFor :: SnapshotNumber}
   | WaitOnSeenSnapshot
+  | WaitOnSeenTxs {unseenTxId :: [TxIdType tx]}
```
Could: name the field in plural as it refers to multiple transaction ids

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -691,11 +703,19 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxs =
+  waitSeenTxs continue =
+    if fromList requestedTxIds `Set.isSubsetOf` fromList (txId <$> seenTxs)
+      then continue
+      else Wait $ WaitOnSeenTxs $ toList $ fromList requestedTxIds `Set.difference` fromList (txId <$> seenTxs)
```
Could: change the type to a `Set (TxIdType tx)` and not convert `toList` here.

Could: use `\\`

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -691,11 +703,19 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxs =
+      else Wait $ WaitOnSeenTxs $ toList $ fromList requestedTxIds `Set.difference` fromList (txId <$> seenTxs)
+
+  resolvedTxs =
+    mapMaybe (\txid -> find (\tx -> txId tx == txid) seenTxs) requestedTxIds
```
This is a bit weird, as we _theoretically_ this could drop not-resolved transaction ids.

Could: make this unrepresentable, by combining the `wait` with the resolution and make `waitSeenTxs $ \resolvedTxs ->`

I also combined two statements from the spec in one function [here](https://github.com/input-output-hk/hydra/pull/904/files#diff-f325f10a5757428ff663484ce4cece514a459a08809aa966ecd6d2950f53a67cR674)

ch1bo replied:
Yeah I was not clear. Are you familiar with "Parse don't validate?" .. that's the feeling I had here and fail to explain it.

ch1bo replied:
Makes sense that this would be an `Error`

ch1bo commented on CHANGELOG.md:
```diff
@@ -23,7 +23,10 @@ following procedure to upgrade all the nodes running a head:
 
-- Fuel removal: 
+- Protocol change: Ensure all transactions requested in a snapshot
+  have been seen before acknowledging it
```
Should: update the changelog and mark this a `**BREAKING**` change.

ch1bo commented on spec/fig_offchain_prot.tex:
```diff
@@ -93,7 +93,7 @@
 
 							% Wait for snapshot no snapshot in flight anymore and all txs applicable
-							\Wait{$\bars = \hats ~ \land ~ \barmU \applytx \mT_{res} \not= \bot$}{ %
+							\Wait{$\bars = \hats ~ \land ~ \barmU \applytx \mT_{res} \not= \bot ~ \and ~ \mT_{res} \subseteq \hatmT$}{ %
```
This would not typecheck if we would have written all this in agda :D

I think we need to change the signature of `onReq` to have ${\mathcal T}_{req}^{\#}$ as argument, then have an additional `wait` with $\forall h \in {\mathcal T}_{req}^{\#} : (h, \cdot) \in {\mathcal T}$ and a resolve statement defining ${\mathcal T}_{res}$, i.e. what I had done here (with $\hat {\mathcal T}$ this time though):

https://github.com/input-output-hk/hydra/issues/728#issuecomment-1460156847

ch1bo replied:
I'd be happy to pair on this.. and then we see together that we should have written the spec in agda

ch1bo replied:
Yeah so the spec has not been updated in this PR? Do we have a follow-up item which makes us not forget this?

ch1bo reviewed (commented):
Wow this was quick! This only requires updating the spec now IMO (more than the one section I commented)

abailly committed:
Pluralise field name with several things

abailly committed:
Adapt logic of ReqSn handling to require txs to be applicable

We previously waited on the transactions in the snapshot to be
applicable to the previous snapshot (confirmed UTxO), but as we now
wait for all tx to have been seen, and seeing a tx implies it to be
applicable to the latest seenUTxO which itself is a derivation from
the previous snapshot, it should be an error to request a snapshot
which does not apply cleanly on the previous snapshot.

pgrange commented on hydra-cardano-api/src/Hydra/Cardano/Api/TxId.hs:
```diff
@@ -1,10 +1,24 @@
 import qualified Cardano.Ledger.TxIn as Ledger
+import Cardano.Binary (ToCBOR(..), FromCBOR (..))
+
+-- missing CBOR instances
+
```
```suggestion
```

pgrange reviewed (commented):
There is the case of double spending that will hang the head forever with this version of the code, I think.

```mermaid
sequenceDiagram
    actor Alice
    actor Bob
    Bob->>Alice: Ta spends u
    Bob->>Alice: Tb spends u
    Note over Alice: Alice reject this double spent
    Bob->>Alice: ReqSn [Tb]
    Note over Alice: Alice stuck forever waiting to see Tb it has rejected before
```

See work from @ch1bo in https://github.com/input-output-hk/hydra/commit/9d19ef4d5860537506b07f8e8bcf8deb0d7062d5

See also https://github.com/input-output-hk/hydra/issues/728#issuecomment-1460156847

pgrange committed:
Test showing the head is stuck on conflicting transaction

pgrange committed:
Store all transactions so that we don't miss conflicting ones

abailly committed:
Add dedicated predicate for wait outcome

pgrange committed:
Generalize combining various HeadLogic Outcomes

To prepare for #903 we want to be able to combine outcomes
of type `Wait` and `NewState`.

This commit allows for that and provide a more cohesive
Outcome algebra.

pgrange committed:
Store all transactions so that we don't miss conflicting ones

abailly committed:
Improve error reporting upon multisig failures

abailly committed:
Failing test checking tx in snapshot are removed from map

abailly committed:
Garbage collect transactions from an accepted snapshot

abailly committed:
Only cleanup txs map when snapshot fully confirmed

abailly committed:
Update specification

abailly committed:
Fix golden files

v0d1ch commented on hydra-node/test/Hydra/CryptoSpec.hs:
```diff
@@ -68,11 +70,18 @@ specMultiSignature =
-       in verifyMultiSignature vks msig msg
+       in verifyMultiSignature vks msig msg === Verified
+
+    prop "aggregateInOrder/verifyMultiSignature roundtrip" $ \sks (msg :: ByteString) ->
```
We probably want to test also what happens when signatures are not verified.

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -613,16 +644,19 @@ onOpenNetworkReqTx ::
       | ttl <= 0 ->
           Effects [ClientEffect $ TxInvalid headId seenUTxO tx err]
-      | otherwise -> Wait $ WaitOnNotApplicableTx err
+      | otherwise ->
```
We should clean all txs when transaction becomes invalid.

v0d1ch committed:
Add a test to assure non applicable tx is removed from the state

- When trying to apply a tx upon ReqTx, if the tx is not applicable and
TTL has passed we are removing the transaction from the known hydra
state.
